Add a design doc for personal cloud accounts

omgjlk opened the pull request:
Signed-off-by: Jesse Keating <jkeating@j2solutions.net>

omgjlk committed:
Add a design doc for personal cloud accounts

Signed-off-by: Jesse Keating <jkeating@j2solutions.net>

jamielennox commented on lore/developers/contributing/designs/personal-cloud-accounts.md:
```diff
@@ -0,0 +1,32 @@
+
+### Accounts Projects and Roles
+
+An account is a set of credentials to log into OpenStack via keystone. An account by itself is not very useful, as it requires a role within a project. A single account may have multiple roles across mutiple projects. For our cloud, each developer is entitled to an account with a matching project, and a role of _member_ within the project.
```
why wouldn't we have project_admin on our own project?

omgjlk replied:
`project_admin` role allows creating new accounts (though they can only be granted roles on the given project). Tracking which of these accounts "belong" to a given project and should be cleaned up should a project get removed would be difficult. I feel it would be best to keep account creation to automation. 

The help doc that is linked to describes the capabilities of the roles.

rattboi commented on lore/developers/contributing/designs/personal-cloud-accounts.md:
```diff
@@ -0,0 +1,32 @@
+
+### Accounts Projects and Roles
+
+An account is a set of credentials to log into OpenStack via keystone. An account by itself is not very useful, as it requires a role within a project. A single account may have multiple roles across mutiple projects. For our cloud, each developer is entitled to an account with a matching project, and a role of _member_ within the project.
```
s/mutiple/multiple/

omgjlk replied:
Fixed.

ConsoleCatzirl commented on lore/developers/contributing/designs/personal-cloud-accounts.md:
```diff
@@ -0,0 +1,32 @@
+
+Each BonnyCI developer is entitled to a personal cloud account and matching project, with an appropriate role. A reasonable set of quotas will be enforced. Accounts and projects will be created via automation.
+
+### Accounts Projects and Roles
```
(nit) I would like to see some commas here: Accounts, Projects, and Roles

ConsoleCatzirl commented on lore/developers/contributing/designs/personal-cloud-accounts.md:
```diff
@@ -0,0 +1,32 @@
+
+### Accounts Projects and Roles
+
+An account is a set of credentials to log into OpenStack via keystone. An account by itself is not very useful, as it requires a role within a project. A single account may have multiple roles across multiple projects. For our cloud, each developer is entitled to an account with a matching project, and a role of _member_ within the project.
```
The same account will be used for both user project and the production project, right? If so, we should mention that production role here as well.

omgjlk replied:
Well, I hadn't really considered granting a personal account a role within the production account(s). The production account(s), and networks attached, are supposed to be part of a different proposal. I'll consider this for a bit.

SpamapS replied:
The blurb in the next commit explains nicely how this works.

loafyloaf commented on lore/developers/contributing/designs/personal-cloud-accounts.md:
```diff
@@ -0,0 +1,32 @@
+
+### Quotas
+
+Quotas are an artificial limit placed on certain consumable resources within an OpenStack cloud. We will use quotas to prevent any one user from consuming all of the resources of our cloud, preventing others from working and preventing our production from having capacity to test things. The quotas stated here are just a default. If a developer has a legitimate need for additional capacity, their  ...
```
being as our (new) cloud has a definite size, would it not make sense to specify the quotas and/or resources alloted to a specific user?

omgjlk replied:
The default quotas are detailed in the link provided. I felt it was better to link than to replicate all the content. If we decide we need to deviate from the defaults, then we could outline them in a the documentation.

This doc is meant to drive the development and dev/user facing documentation. It is not the documentation itself.

omgjlk committed:
Add blurb about production projects

Add a little bit about how personal accounts will also be granted rights
on production projects.

Signed-off-by: Jesse Keating <jkeating@j2solutions.net>

missaugustina commented on lore/developers/contributing/designs/personal-cloud-accounts.md:
```diff
@@ -0,0 +1,34 @@
+
+We will create and manage accounts, projects, roles, and quotas via Ansible automation. Due to these actions requiring a `cloud_admin` level role, the automation will be ran by a human with the appropriate credentials to start with.
+
+While most developers will have access to `cloud_admin` level credentials, it is critical to not abuse those credentials to manage accounts, projects, roles, or quotas outside of the automation.
```
We should probably make some statement that we are currently determining our user agreement criteria that defines what constitutes abuse.

What is the process for notifying users about abuse? What would be the process for users that we determine are abusing our resources? Maybe something along the lines of notifying them and talking to them about their use case to see if we can figure out a better workflow. If the user fails to cooperate with what we've agreed then we would disable their account/services?

SpamapS replied:
I believe the sentence defines what abuse would be: "manage accounts, projects, roles, or quotas outside of the automation."

omgjlk replied:
Seems like I could have made it more clear that this design and policy, and the accounts themselves are for the core BonnyCI (the service) developers, not for BonnyCI consumers. Perhaps a blurb about that, plus explaining that account requests can be made via pull requests to the automation would clarify.

missaugustina reviewed (changes requested):
I'm not sure of the exact scope of these accounts but if these are at all user facing, we should probably see about drafting some kind of user agreement that makes it clear what our expectations are for how they will use our service and to cover our asses in case it turns out someone is using our services for anything that could put us on the hook legally (eg, spamming, child porn, warez). If this is outside of the scope of this then a clarifying statement that these accounts are not user-facing per se and they are only accessible and managed by the BonnyCI team or something.

gandelman-a commented:
This LGTM but will wait until other people's requests have been addressed.

SpamapS reviewed (commented):
No changes needed IMO. Thanks @j2sol , +1

omgjlk commented:
@missaugustina This design doc is meant to drive the development of the automation, and the creation of documentation. It is not meant to be the documentation itself that devs would refer to in order to determine how to get an account or whatnot. It's just to show the decision process that led to the automation in the first place.

The accounts themselves are meant for the core BonnyCI dev team, not the end users of the BonnyCI service. I'll add a blurb that clarifies this somewhere.

omgjlk committed:
Clarify that accounts are for devs, not users

Add clarification about who gets accounts, and how they would be
requested.

Signed-off-by: Jesse Keating <jkeating@j2solutions.net>
